src/mds/Mutation.h

```diff
@@ -389,6 +389,30 @@ struct MDRequestImpl : public MutationImpl {
   bool is_queued_for_replay() const;
   int compare_paths();
 
+  // SnapDiff binds synthetic (aka effective) snapids to returned returns.
+  // Such an id is a union of the first and last snapshots' ids. Actually -
+  // their lower 32-bits. This is necessary to get unique snapids for
+  // different SnapDiff requests's results and avoid a mess at clients.
+  // MDS should be able to retrieve the real (aka primary) object snapshot id
+  // from the effective one to access objects properly while processing
+  // subsequent requests on SnapDiff entries.
+  // For new/updated entries such a primary id is the last snapshot id, while
+  // removed ones use the first snapshot id for that.
+  // Primary snapshot id is encoded as a lower 32-bits of the effective id.
+  // And higher 32-bits keep the secondary (either first or last snapid for
+  // updated or removed entries respectively).
+  // The following function properly builds effective snapid as per above.
```

Comment:
This needs some examples perhaps and more explanation. I confess I'm mystified by this comment.

Reply:
>I think better would be to create an API for building a snapdiff stream. Something like:

>What do you think?
I refactored this part a bit, please take a look.